tests/test_butterworth_filter.py

```diff
@@ -0,0 +1,192 @@
+    with pytest.raises(ValueError):
+        ButterworthFilter(**filter_arguments)
+
+
+def test_butterworth_cutoff_frequency(filter_arguments):
+    del filter_arguments["cutoff_frequency"]
+
+    with pytest.raises(TypeError):
+        ButterworthFilter(**filter_arguments, cutoff_frequency="not a number")
+
+    filter_arguments["filter_type"] = "bandpass"
+    with pytest.raises(TypeError):
+        ButterworthFilter(**filter_arguments, cutoff_frequency="not a number")
+
+    with pytest.raises(TypeError):
+        ButterworthFilter(**filter_arguments, cutoff_frequency=10)
+
+    with pytest.raises(TypeError):
+        ButterworthFilter(**filter_arguments, cutoff_frequency=("not a number", True))
+
+    with pytest.raises(TypeError):
+        ButterworthFilter(**filter_arguments, cutoff_frequency=[20, 30])
+
+    with pytest.raises(ValueError):
+        ButterworthFilter(**filter_arguments, cutoff_frequency=(1,))
+
+    with pytest.raises(ValueError):
+        ButterworthFilter(**filter_arguments, cutoff_frequency=(1, 2, 3))
+
+    try:
+        ButterworthFilter(**filter_arguments, cutoff_frequency=(20, 30))
+    except (ValueError, TypeError):
+        pytest.fail("Unexpected error")
+
+
+def test_butterworth_sample_frequency(filter_arguments):
+    del filter_arguments["sample_frequency"]
+
+    with pytest.raises(TypeError):
+        ButterworthFilter(**filter_arguments, sample_frequency="a string")
+
+    with pytest.raises(ValueError):
+        ButterworthFilter(**filter_arguments, sample_frequency=-1)
+
+    try:
+        ButterworthFilter(**filter_arguments, sample_frequency=1)
+    except (TypeError, ValueError):
+        pytest.fail("Unexpected error")
+
+
+def test_create_specified_filter(filter_arguments):
+    lp_filter = LowPassFilter(**filter_arguments)
+
+    with pytest.raises(AttributeError):
+        hp_filter = HighPassFilter(**filter_arguments)
+
+    del filter_arguments["filter_type"]
+
+    lp_filter = LowPassFilter(**filter_arguments)
+    hp_filter = HighPassFilter(**filter_arguments)
+
+    for filter_ in (lp_filter, hp_filter):
+        check_filter_attributes(filter_, filter_arguments)
+
+    filter_arguments["cutoff_frequency"] = (20, 30)
+    bp_filter = BandPassFilter(**filter_arguments)
+    bs_filter = BandStopFilter(**filter_arguments)
+
+    for filter_ in (bp_filter, bs_filter):
+        check_filter_attributes(filter_, filter_arguments)
+
+    assert lp_filter.filter_type == "lowpass"
+    assert hp_filter.filter_type == "highpass"
+    assert bp_filter.filter_type == "bandpass"
+    assert bs_filter.filter_type == "bandstop"
+
+
+def test_specified_butterworth_equivalence(filter_arguments):
+    del filter_arguments["filter_type"]
+
+    filter1 = ButterworthFilter(**filter_arguments, filter_type="lowpass")
+    filter2 = LowPassFilter(**filter_arguments)
+
+    # filter1.__eq__(filter2) differs from filter2.__eq__(filter1)
+    assert filter1 == filter2
+    assert filter2 == filter1
```

Comment:
what is the difference?

Reply:
In an early version of the `__eq__()` method I wrote (now removed and replaced by the dataclass version) where both lines would function differently when the filters were of different types (one ButterworthFilter, one a subclass). This difference is no longer present. The second line is removed in d576a82c5ebbc064c224f1839ec91351bde31f25.